src/compiler/checker.ts

```diff
@@ -5595,9 +5595,14 @@ namespace ts {
 
                 if (source.flags & TypeFlags.TypeParameter) {
                     let constraint = getConstraintOfTypeParameter(<TypeParameter>source);
+
                     if (!constraint || constraint.flags & TypeFlags.Any) {
                         constraint = emptyObjectType;
                     }
+
+                    // The constraint may need to be further instantiated with its 'this' type.
+                    constraint = getTypeWithThisArgument(constraint);

```

Comment:
Let me check my understanding: an unconstrained type parameter gets 'any' or '{}' as its constraint, which shouldn't have a this-type, so this is a no-op there. Am I right?